File: src/main/java/com/yrhp/crud/util/SlugUtil.java
package com.yrhp.crud.util;

import java.text.Normalizer;
import java.util.regex.Pattern;

/**
 * Utility class for generating and handling URL-friendly slugs from company names.
 *
 * Slug generation rules:
 * - Convert to lowercase
 * - Replace spaces with hyphens
 * - Remove special characters, keeping only alphanumeric and hyphens
 * - Remove consecutive hyphens
 *
 * Examples:
 * "YRHP Pvt Ltd" -> "yrhp-pvt-ltd"
 * "ABC & Co. Ltd" -> "abc-co-ltd"
 * "Tech-Solutions 2024" -> "tech-solutions-2024"
 */
public class SlugUtil {

    private static final Pattern NON_ALPHANUMERIC = Pattern.compile("[^a-z0-9-]");
    private static final Pattern MULTIPLE_HYPHENS = Pattern.compile("-+");

    /**
     * Generates a URL-friendly slug from the given company name.
     *
     * @param companyName the company name to convert
     * @return a URL-friendly slug (lowercase, hyphens, alphanumeric only)
     */
    public static String generateSlug(String companyName) {
        if (companyName == null || companyName.trim().isEmpty()) {
            throw new IllegalArgumentException("Company name cannot be null or empty");
        }

        // Step 1: Normalize unicode characters (e.g., accents)
        String normalized = Normalizer.normalize(companyName, Normalizer.Form.NFD);

        // Step 2: Convert to lowercase
        String lowercase = normalized.toLowerCase();

        // Step 3: Replace spaces with hyphens
        String withHyphens = lowercase.replaceAll("\\s+", "-");

        // Step 4: Remove special characters except hyphens and alphanumeric
        String cleaned = NON_ALPHANUMERIC.matcher(withHyphens).replaceAll("");

        // Step 5: Replace multiple consecutive hyphens with single hyphen
        String slug = MULTIPLE_HYPHENS.matcher(cleaned).replaceAll("-");

        // Step 6: Remove leading/trailing hyphens
        slug = slug.replaceAll("^-+|-+$", "");

        if (slug.isEmpty()) {
            throw new IllegalArgumentException("Generated slug is empty for company name: " + companyName);
        }

        return slug;
    }

    /**
     * Checks if the provided string is a valid slug.
     * A valid slug contains only lowercase alphanumeric characters and hyphens,
     * and doesn't start or end with a hyphen.
     *
     * @param slug the slug to validate
     * @return true if the slug is valid, false otherwise
     */
    public static boolean isValidSlug(String slug) {
        if (slug == null || slug.isEmpty()) {
            return false;
        }
        return slug.matches("^[a-z0-9]+(-[a-z0-9]+)*$");
    }
}

File: review-generator-frontend/src/utils/slugUtil.js
/**
 * Slug utility for the frontend
 * Handles slug generation and validation
 * Note: Primary slug generation happens on the backend (SlugUtil.java)
 * This is used for validation and reference purposes only
 */

/**
 * Generates a URL-friendly slug from a company name
 * This matches the backend SlugUtil behavior
 *
 * @param {string} companyName - The company name to convert
 * @returns {string} A URL-friendly slug (lowercase, hyphens, alphanumeric only)
 */
export const generateSlug = (companyName) => {
  if (!companyName || companyName.trim() === '') {
    throw new Error('Company name cannot be null or empty');
  }

  // Convert to lowercase
  let slug = companyName.toLowerCase();

  // Replace spaces with hyphens
  slug = slug.replace(/\s+/g, '-');

  // Remove special characters except hyphens and alphanumeric
  slug = slug.replace(/[^a-z0-9-]/g, '');

  // Replace multiple consecutive hyphens with single hyphen
  slug = slug.replace(/-+/g, '-');

  // Remove leading/trailing hyphens
  slug = slug.replace(/^-+|-+$/g, '');

  if (slug === '') {
    throw new Error(`Generated slug is empty for company name: ${companyName}`);
  }

  return slug;
};

/**
 * Validates if a slug is properly formatted
 * A valid slug contains only lowercase alphanumeric characters and hyphens,
 * and doesn't start or end with a hyphen
 *
 * @param {string} slug - The slug to validate
 * @returns {boolean} True if valid, false otherwise
 */
export const isValidSlug = (slug) => {
  if (!slug || slug === '') {
    return false;
  }
  return /^[a-z0-9]+(-[a-z0-9]+)*$/.test(slug);
};

/**
 * Normalizes a slug by applying standard slug transformation
 * Useful for comparing slugs or standardizing input
 *
 * @param {string} slug - The slug to normalize
 * @returns {string} The normalized slug
 */
export const normalizeSlug = (slug) => {
  if (!slug) return '';
  return slug.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '').replace(/-+/g, '-').replace(/^-+|-+$/g, '');
};
